File: examples/java-junit-maven/src/test/java/examples/Library.java
package examples;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.List;

public final class Library {

    /** An amount in a single currency; the value is in the currency's major unit (pounds). */
    public record Money(String currency, double value) {}

    public static Money gbp(double value) {
        return new Money("GBP", value);
    }

    public static final Money FEE_PER_DAY = gbp(0.5);

    public record Loan(String title, LocalDate due) {}

    private Library() {}

    public static Money addMoney(Money a, Money b) {
        if (!a.currency().equals(b.currency())) {
            throw new IllegalArgumentException("cannot add " + b.currency() + " to " + a.currency());
        }
        return new Money(a.currency(), a.value() + b.value());
    }

    /** Fee for returning a loan: 50p per day past the due date. */
    public static Money lateFee(Loan loan, LocalDate returnedOn) {
        long daysLate = Math.max(0, ChronoUnit.DAYS.between(loan.due(), returnedOn));
        return gbp(daysLate * FEE_PER_DAY.value());
    }

    /** A member may borrow as long as none of their loans is overdue. */
    public static boolean mayBorrow(List<Loan> loans, LocalDate on) {
        return loans.stream().noneMatch(loan -> loan.due().isBefore(on));
    }
}

File: examples/typescript-vitest/src/library.ts
export type Money = {
  readonly currency: string
  readonly value: number
}

export const GBP = (value: number): Money => ({ currency: 'GBP', value })

export const FEE_PER_DAY = GBP(0.5)

export function addMoney(a: Money, b: Money): Money {
  if (a.currency !== b.currency) throw new Error(`cannot add ${b.currency} to ${a.currency}`)
  return { currency: a.currency, value: a.value + b.value }
}

export type Loan = {
  readonly title: string
  readonly due: Date
}

const MS_PER_DAY = 24 * 60 * 60 * 1000

export function lateFee(loan: Loan, returnedOn: Date): Money {
  const daysLate = Math.max(0, (returnedOn.getTime() - loan.due.getTime()) / MS_PER_DAY)
  return GBP(daysLate * FEE_PER_DAY.value)
}

export function mayBorrow(loans: ReadonlyArray<Loan>, on: Date): boolean {
  return loans.every((loan) => loan.due.getTime() >= on.getTime())
}
